src/ops/fuelup_self.rs

```diff
@@ -1,17 +1,78 @@
-use anyhow::Result;
+use anyhow::{bail, Result};
+use std::{fs, path::Path};
+use tempfile::tempdir_in;
+use tracing::{error, info};
 
 use crate::{
     component,
     download::{download_file_and_unpack, unpack_bins, DownloadCfg},
-    path::fuelup_bin_dir,
+    path::{fuelup_bin, fuelup_bin_dir},
 };
 
+pub fn attempt_install_self(download_cfg: DownloadCfg, dst: &Path) -> Result<()> {
+    download_file_and_unpack(&download_cfg, dst)?;
+    unpack_bins(dst, dst)?;
+
+    Ok(())
+}
+
 pub fn self_update() -> Result<()> {
     let download_cfg = DownloadCfg::new(component::FUELUP, None)?;
+
+    let fuelup_bin = fuelup_bin();
     let fuelup_bin_dir = fuelup_bin_dir();
+    let fuelup_backup = fuelup_bin_dir.join("fuelup-backup");
+
+    let fuelup_new_dir = tempdir_in(&fuelup_bin_dir)?;
+    let fuelup_new_dir_path = fuelup_new_dir.path();
+    let fuelup_new = fuelup_new_dir_path.join(component::FUELUP);
+
+    if let Err(e) = attempt_install_self(download_cfg, fuelup_new_dir_path) {
+        // Skip all other steps if downloading fails here.
+        // We do not need to handle failure, since this downloads to a tempdir.
+        bail!("Failed to install fuelup: {}", e);
+    };
+
+    if fuelup_bin.exists() {
+        // Make a backup of fuelup, fuelup-backup.
+        if let Err(e) = fs::rename(&fuelup_bin, &fuelup_backup) {
+            bail!(
+                "Failed moving {} to {}: {}",
+                &fuelup_bin.display(),
+                &fuelup_backup.display(),
+                e
+            );
+        }
+    };
+
+    info!(
+        "Moving {} to {}",
+        fuelup_new.display(),
+        &fuelup_bin.display()
+    );
+    if let Err(e) = fs::rename(&fuelup_new, &fuelup_bin) {
+        error!(
+            "Failed to replace old fuelup with new fuelup: {}. Attempting to restore backup fuelup.",
+        e);
```

Comment:
This is such strange formatting, I've never seen this from rustfmt :laughing: That said if rustfmt's happy, I'm happy.